diff --git a/ClassIsland.Shared/Models/Profile/ClassInfo.cs b/ClassIsland.Shared/Models/Profile/ClassInfo.cs
index ccb26898..2138f066 100644
--- a/ClassIsland.Shared/Models/Profile/ClassInfo.cs
+++ b/ClassIsland.Shared/Models/Profile/ClassInfo.cs
@@ -50,7 +50,18 @@ public class ClassInfo : AttachableSettingsObject
     /// <summary>
     /// 课程对应的时间点
     /// </summary>
-    [JsonIgnore] public TimeLayoutItem CurrentTimeLayoutItem => (from i in CurrentTimeLayout.Layouts where i.TimeType == 0 select i).ToList()[Index];
+    [JsonIgnore] public TimeLayoutItem CurrentTimeLayoutItem
+    {
+        get
+        {
+            var valid = (from i in CurrentTimeLayout.Layouts where i.TimeType == 0 select i).ToList();
+            if (Index < 0 || Index >= valid.Count)
+            {
+                return TimeLayoutItem.Empty;
+            }
+            return valid[Index];
+        }
+    }
 
 
     /// <summary>
