diff --git a/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs b/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs
index 4f94092c..3403e6f6 100644
--- a/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs
+++ b/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs
@@ -38,7 +38,7 @@ public class AfterSchoolNotificationProvider : NotificationProviderBase<AfterSch
     {
         var settings = (IAfterSchoolNotificationProviderSettingsBase?)GetAttachedSettings() ?? Settings;
         var now = ExactTimeService.GetCurrentLocalDateTime().TimeOfDay;
-        var afterSchoolTime = LessonsService.CurrentClassPlan?.TimeLayout.Layouts.LastOrDefault(i => i.TimeType is 0 or 1)?.EndSecond.TimeOfDay;
+        var afterSchoolTime = LessonsService.ValidTimeLayoutItems.LastOrDefault(i => i.TimeType is 0 or 1)?.EndSecond.TimeOfDay;
         if (!settings.IsEnabled ||
             (now - afterSchoolTime) > TimeSpan.FromSeconds(10))
             return;
