diff --git a/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.cs b/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.cs
index 1dd5baae..d65d16e6 100644
--- a/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.cs
+++ b/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.cs
@@ -117,11 +117,6 @@ public class LessonsListBox : ListBox
         set { SetValue(HideFinishedClassProperty, value); }
     }
 
-    /// <summary>
-    /// 课程服务
-    /// </summary>
-    public ILessonsService? LessonsService => IAppHost.TryGetService<ILessonsService>();
-
     static LessonsListBox()
     {
         DefaultStyleKeyProperty.OverrideMetadata(typeof(LessonsListBox), new FrameworkPropertyMetadata(typeof(LessonsListBox)));
diff --git a/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.xaml b/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.xaml
index 21426e7f..c52f2db1 100644
--- a/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.xaml
+++ b/ClassIsland.Core/Controls/LessonsControls/LessonsListBox.xaml
@@ -99,8 +99,7 @@
                                 <Binding Path="DiscardHidingDefault" RelativeSource="{RelativeSource FindAncestor, AncestorType=lessonsControls:LessonsListBox}" Mode="OneWay"/>
                                 <Binding Path="ShowCurrentTimeLayoutItemOnlyOnClass" RelativeSource="{RelativeSource FindAncestor, AncestorType=lessonsControls:LessonsListBox}" Mode="OneWay"/>
                                 <Binding Path="HideFinishedClass" RelativeSource="{RelativeSource FindAncestor, AncestorType=lessonsControls:LessonsListBox}" Mode="OneWay"/>
-                                <Binding Path="LessonsService.ValidTimeLayoutItems" RelativeSource="{RelativeSource FindAncestor, AncestorType=lessonsControls:LessonsListBox}" Mode="OneWay"/>
-                                <Binding Path="IsLiveUpdatingEnabled" RelativeSource="{RelativeSource FindAncestor, AncestorType=lessonsControls:LessonsListBox}" Mode="OneWay"/>
+                                <Binding Path="ClassPlan.ValidTimeLayoutItems" RelativeSource="{RelativeSource FindAncestor, AncestorType=lessonsControls:LessonsListBox}" Mode="OneWay"/>
                             </MultiBinding>
                         </ContentPresenter.ContentTemplate>
                     </ContentPresenter>
diff --git a/ClassIsland.Core/Controls/LessonsControls/LessonsListBoxItemTemplateMultiConverter.cs b/ClassIsland.Core/Controls/LessonsControls/LessonsListBoxItemTemplateMultiConverter.cs
index a987dbc6..9023ba1b 100644
--- a/ClassIsland.Core/Controls/LessonsControls/LessonsListBoxItemTemplateMultiConverter.cs
+++ b/ClassIsland.Core/Controls/LessonsControls/LessonsListBoxItemTemplateMultiConverter.cs
@@ -51,14 +51,13 @@ internal class LessonsListBoxItemTemplateMultiConverter : DependencyObject, IMul
         // [6]: bool             HideFinishedClass
         // [7]: ICollection<...> ValidTimePoints
         // [8]: bool             IsLiveUpdatingEnabled
-        if (values.Length < 9)
+        if (values.Length < 8)
             return BlankDataTemplate;
         if (values[0] is not int timeType ||
             values[1] is not bool isHideDefault ||
             values[4] is not bool discardHidingDefault ||
             values[5] is not bool showCurrentTimeLayoutItemOnlyOnClass ||
-            values[6] is not bool hideFinishedClass ||
-            values[8] is not bool isLiveUpdateEnabled)
+            values[6] is not bool hideFinishedClass)
         {
             return BlankDataTemplate;
         }
@@ -73,7 +72,7 @@ internal class LessonsListBoxItemTemplateMultiConverter : DependencyObject, IMul
         if (currentItem != selectedItem && selectedItem?.TimeType == 0 && showCurrentTimeLayoutItemOnlyOnClass)
             return BlankDataTemplate;
 
-        if (isLiveUpdateEnabled && values[7] is ICollection<TimeLayoutItem> validTimePoints &&
+        if (values[7] is ICollection<TimeLayoutItem> validTimePoints &&
             (currentItem == null || !validTimePoints.Contains(currentItem)))
         {
             return BlankDataTemplate;
diff --git a/ClassIsland.Shared.IPC/Abstractions/Services/IPublicLessonsService.cs b/ClassIsland.Shared.IPC/Abstractions/Services/IPublicLessonsService.cs
index 93621045..bfa15772 100644
--- a/ClassIsland.Shared.IPC/Abstractions/Services/IPublicLessonsService.cs
+++ b/ClassIsland.Shared.IPC/Abstractions/Services/IPublicLessonsService.cs
@@ -102,9 +102,4 @@ public interface IPublicLessonsService
     /// <param name="date">要获取课表的日期</param>
     /// <returns>获取到的课表</returns>
     ClassPlan? GetClassPlanByDate(DateTime date);
-
-    /// <summary>
-    /// 有效的时间点列表。
-    /// </summary>
-    ObservableCollection<TimeLayoutItem> ValidTimeLayoutItems { get; set; }
 }
\ No newline at end of file
diff --git a/ClassIsland.Shared/Models/Profile/ClassPlan.cs b/ClassIsland.Shared/Models/Profile/ClassPlan.cs
index f62f25fb..bacd18d8 100644
--- a/ClassIsland.Shared/Models/Profile/ClassPlan.cs
+++ b/ClassIsland.Shared/Models/Profile/ClassPlan.cs
@@ -24,6 +24,77 @@ public class ClassPlan : AttachableSettingsObject
     private int _lastTimeLayoutCount = -1;
     private string _associatedGroup = ClassPlanGroup.DefaultGroupGuid.ToString();
 
+    private bool _isValidTimeLayoutItemsDirty = true;
+    private ObservableCollection<TimeLayoutItem> _validTimeLayoutItems = [];
+
+    /// <summary>
+    /// 有效的时间点列表
+    /// </summary>
+    [JsonIgnore]
+    public ObservableCollection<TimeLayoutItem> ValidTimeLayoutItems
+    {
+        get
+        {
+            if (!_isValidTimeLayoutItemsDirty) 
+                return _validTimeLayoutItems;
+            _validTimeLayoutItems = GetValidTimeLayoutItems();
+            _isValidTimeLayoutItemsDirty = false;
+            return _validTimeLayoutItems;
+        }
+    }
+
+    private void MakeValidTimeLayoutItemsDirty()
+    {
+        _isValidTimeLayoutItemsDirty = true;
+        OnPropertyChanged(nameof(ValidTimeLayoutItems));
+    }
+
+    private ObservableCollection<TimeLayoutItem> GetValidTimeLayoutItems()
+    {
+        if (TimeLayout == null)
+            return [];
+        var timeLayoutMap = Classes.ToDictionary(x => x.CurrentTimeLayoutItem, x => x);
+        var displayTimePoints = TimeLayout.Layouts
+            .Where(x => x.TimeType is 0 or 1 or 2)
+            .ToList();
+        ObservableCollection<TimeLayoutItem> items = [.. displayTimePoints.Select(x => x)];
+        List<TimeLayoutItem> remove = [];
+        // 正向搜索
+        var isPrevEnabled = true;
+        for (var i = 0; i < displayTimePoints.Count; i++)
+        {
+            if (timeLayoutMap.TryGetValue(items[i], out var info))
+            {
+                isPrevEnabled = info.IsEnabled;
+            }
+
+            if (!isPrevEnabled)
+            {
+                remove.Add(displayTimePoints[i]);
+            }
+        }
+        // 反向搜索
+        isPrevEnabled = true;
+        for (var i = displayTimePoints.Count - 1; i >= 0; i--)
+        {
+            if (timeLayoutMap.TryGetValue(items[i], out var info))
+            {
+                isPrevEnabled = info.IsEnabled;
+            }
+
+            if (!isPrevEnabled)
+            {
+                remove.Add(displayTimePoints[i]);
+            }
+        }
+
+        foreach (var i in remove)
+        {
+            items.Remove(i);
+        }
+        return items;
+    }
+
     /// <summary>
     /// 当课程表更新时触发
     /// </summary>
@@ -60,6 +131,7 @@ public class ClassPlan : AttachableSettingsObject
             TimeLayout.LayoutItemChanged += TimeLayoutOnLayoutItemChanged;
         }
         Classes.CollectionChanged += ClassesOnCollectionChanged;
+        ClassesChanged += (sender, args) => MakeValidTimeLayoutItemsDirty();
     }
 
     private void ClassesOnCollectionChanged(object? sender, NotifyCollectionChangedEventArgs e)
diff --git a/ClassIsland/Services/LessonsService.cs b/ClassIsland/Services/LessonsService.cs
index 1f806486..5ab0f7d1 100644
--- a/ClassIsland/Services/LessonsService.cs
+++ b/ClassIsland/Services/LessonsService.cs
@@ -35,7 +35,6 @@ public class LessonsService : ObservableRecipient, ILessonsService
     private TimeSpan _onBreakingTimeLeftTime = TimeSpan.Zero;
     private TimeLayoutItem _nextClassTimeLayoutItem = TimeLayoutItem.Empty;
     private ObservableCollection<int> _multiWeekRotation = [0, 0, 1, 1, 1];
-    private ObservableCollection<TimeLayoutItem> _validTimeLayoutItems = [];
 
     private static readonly ObservableCollection<int> DefaultMultiWeekRotation = [0, 0, 1, 1, 1];
 
@@ -148,58 +147,6 @@ public class LessonsService : ObservableRecipient, ILessonsService
 
     public ClassPlan? GetClassPlanByDate(DateTime date) => GetClassPlanByDate(date, out _);
 
-    public ObservableCollection<TimeLayoutItem> ValidTimeLayoutItems
-    {
-        get => _validTimeLayoutItems;
-        set => SetProperty(ref _validTimeLayoutItems, value);
-    }
-
-    private ObservableCollection<TimeLayoutItem> GetValidTimeLayoutItems()
-    {
-        if (CurrentClassPlan?.TimeLayout == null) 
-            return [];
-        var timeLayoutMap = CurrentClassPlan.Classes.ToDictionary(x => x.CurrentTimeLayoutItem, x => x);
-        var displayTimePoints = CurrentClassPlan.TimeLayout.Layouts
-            .Where(x => x.TimeType is 0 or 1 or 2)
-            .ToList();
-        ObservableCollection<TimeLayoutItem> items = [..displayTimePoints.Select(x => x)];
-        List<TimeLayoutItem> remove = [];
-        // 正向搜索
-        var isPrevEnabled = true;
-        for (var i = 0; i < displayTimePoints.Count; i++)
-        {
-            if (timeLayoutMap.TryGetValue(items[i], out var info))
-            {
-                isPrevEnabled = info.IsEnabled;
-            }
-
-            if (!isPrevEnabled)
-            {
-                remove.Add(displayTimePoints[i]);
-            }
-        }
-        // 反向搜索
-        isPrevEnabled = true;
-        for (var i = displayTimePoints.Count - 1; i >= 0; i--)
-        {
-            if (timeLayoutMap.TryGetValue(items[i], out var info))
-            {
-                isPrevEnabled = info.IsEnabled;
-            }
-
-            if (!isPrevEnabled)
-            {
-                remove.Add(displayTimePoints[i]);
-            }
-        }
-
-        foreach (var i in remove)
-        {
-            items.Remove(i);
-        }
-        return items;
-    }
-
     public ClassPlan? GetClassPlanByDate(DateTime date, out string? guid)
     {
         guid = null;
@@ -336,7 +283,6 @@ public class LessonsService : ObservableRecipient, ILessonsService
         {
             CurrentClassPlan.ClassesChanged += CurrentClassPlanOnClassesChanged;
             CurrentClassPlan.RefreshIsChangedClass();
-            ValidTimeLayoutItems = GetValidTimeLayoutItems();
         }
     }
 
@@ -350,7 +296,7 @@ public class LessonsService : ObservableRecipient, ILessonsService
 
     private void CurrentClassPlanOnClassesChanged(object? sender, EventArgs e)
     {
-        ValidTimeLayoutItems = GetValidTimeLayoutItems();
+        
     }
 
     private bool CurrentSubjectHandler(object? settings)
@@ -487,9 +433,10 @@ public class LessonsService : ObservableRecipient, ILessonsService
         CurrentClassPlan.TimeLayout.IsActivated = true;
 
         var now = ExactTimeService.GetCurrentLocalDateTime().TimeOfDay;
+        var validTimeLayoutItems = CurrentClassPlan.ValidTimeLayoutItems;
 
         // 获取当前时间点信息
-        currentTimeLayoutItem = ValidTimeLayoutItems.FirstOrDefault(i =>
+        currentTimeLayoutItem = validTimeLayoutItems.FirstOrDefault(i =>
             i.TimeType is 0 or 1 &&
             i.StartSecond.TimeOfDay <= now &&
             i.EndSecond.TimeOfDay >= now);
@@ -516,7 +463,7 @@ public class LessonsService : ObservableRecipient, ILessonsService
         }
 
         // 获取下节时间点信息
-        nextClassTimeLayoutItem = ValidTimeLayoutItems.FirstOrDefault(i =>
+        nextClassTimeLayoutItem = validTimeLayoutItems.FirstOrDefault(i =>
             i.TimeType == 0 &&
             i.EndSecond.TimeOfDay >= now);
         if (nextClassTimeLayoutItem != null)
@@ -526,7 +473,7 @@ public class LessonsService : ObservableRecipient, ILessonsService
                 Profile.Subjects.TryGetValue(CurrentClassPlan.Classes[i0].SubjectId, out var subject))
                 nextClassSubject = subject;
         }
-        nextBreakingTimeLayoutItem = ValidTimeLayoutItems.FirstOrDefault(i =>
+        nextBreakingTimeLayoutItem = validTimeLayoutItems.FirstOrDefault(i =>
             i.TimeType == 1 &&
             i.EndSecond.TimeOfDay >= now);
 
diff --git a/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs b/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs
index 3403e6f6..77d24c72 100644
--- a/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs
+++ b/ClassIsland/Services/NotificationProviders/AfterSchoolNotificationProvider.cs
@@ -38,7 +38,7 @@ public class AfterSchoolNotificationProvider : NotificationProviderBase<AfterSch
     {
         var settings = (IAfterSchoolNotificationProviderSettingsBase?)GetAttachedSettings() ?? Settings;
         var now = ExactTimeService.GetCurrentLocalDateTime().TimeOfDay;
-        var afterSchoolTime = LessonsService.ValidTimeLayoutItems.LastOrDefault(i => i.TimeType is 0 or 1)?.EndSecond.TimeOfDay;
+        var afterSchoolTime = LessonsService.CurrentClassPlan?.ValidTimeLayoutItems.LastOrDefault(i => i.TimeType is 0 or 1)?.EndSecond.TimeOfDay;
         if (!settings.IsEnabled ||
             (now - afterSchoolTime) > TimeSpan.FromSeconds(10))
             return;
